lib/graphql/language/printer.rb

```diff
@@ -21,294 +47,344 @@ class Printer
       end
 
       def print_inline_fragment(inline_fragment, indent: "")
-        out = "#{indent}...".dup
+        @out << "#{indent}..."
         if inline_fragment.type
-          out << " on #{print_node(inline_fragment.type)}"
+          @out << " on "
+          print_node(inline_fragment.type)
         end
-        out << print_directives(inline_fragment.directives)
-        out << print_selections(inline_fragment.selections, indent: indent)
-        out
-      end
-
-      def print_input_object(input_object)
-        "{#{input_object.arguments.map { |a| print_argument(a) }.join(", ")}}"
+        print_directives(inline_fragment.directives)
+        print_selections(inline_fragment.selections, indent:)
       end
 
       def print_list_type(list_type)
-        "[#{print_node(list_type.of_type)}]".dup
+        @out << "["
+        print_node(list_type.of_type)
+        @out << "]"
       end
 
       def print_non_null_type(non_null_type)
-        "#{print_node(non_null_type.of_type)}!".dup
+        print_node(non_null_type.of_type)
+        @out << "!"
       end
 
       def print_operation_definition(operation_definition, indent: "")
-        out = "#{indent}#{operation_definition.operation_type}".dup
-        out << " #{operation_definition.name}" if operation_definition.name
+        @out << "#{indent}#{operation_definition.operation_type}"
+        @out << " #{operation_definition.name}" if operation_definition.name
 
         if operation_definition.variables.any?
-          out << "(#{operation_definition.variables.map { |v| print_variable_definition(v) }.join(", ")})"
+          @out << "("
+          operation_definition.variables.each_with_index do |v, i|
+            print_variable_definition(v)
+            @out << ", " if i < operation_definition.variables.size - 1
+          end
+          @out << ")"
         end
 
-        out << print_directives(operation_definition.directives)
-        out << print_selections(operation_definition.selections, indent: indent)
-        out
+        print_directives(operation_definition.directives)
+        print_selections(operation_definition.selections, indent:)
       end
 
       def print_type_name(type_name)
-        "#{type_name.name}".dup
+        @out << type_name.name
       end
 
       def print_variable_definition(variable_definition)
-        out = "$#{variable_definition.name}: #{print_node(variable_definition.type)}".dup
-        out << " = #{print_node(variable_definition.default_value)}" unless variable_definition.default_value.nil?
-        out
+        @out << "$#{variable_definition.name}: "
+        print_node(variable_definition.type)
+        unless variable_definition.default_value.nil?
+          @out << " = "
+          print_node(variable_definition.default_value)
+        end
       end
 
       def print_variable_identifier(variable_identifier)
-        "$#{variable_identifier.name}".dup
+        @out << "$#{variable_identifier.name}"
```

Comment:
Why not go all in...?

```suggestion
        @out << "$" << variable_identifier.name
```

Reply:
There was no reason other than piggy backing off of the existing code. I assume most of the performance gains came from not duplicating strings unnecessarily. Benchmarking the two it does not seem to make a big difference. Interpolations seem to be even faster if there are multiple interpolated variables in the same string. I can make the change though I don't have strong opinions there.

Reply:
I ran a benchmark again with the suggestion with all the interpolations removed and it seems like it's actually a bit slower. I suspect because of the method call indirection `TruncatableBuffer#<<`.

I renamed `#<<` to `#append` to avoid confusion and kept the interpolation.